File: scripts/algorithms/M/Maximize Score After N Operations/Maximize Score After N Operations.py
# Runtime: 6454 ms (Top 9.35%) | Memory: 24.7 MB (Top 37.40%)
from functools import lru_cache

class Solution:
    def maxScore(self, nums: List[int]) -> int:
        def gcd(a, b):
            while a:
                a, b = b%a, a
            return b
        halfplus = len(nums)//2 + 1
        @lru_cache(None)
        def dfs(mask, k):
            if k == halfplus:
                return 0
            res = 0
            for i in range(len(nums)):
                for j in range(i+1, len(nums)):
                    if not(mask & (1<<i)) and not(mask &(1<<j)):
                        res = max(res, k*gcd(nums[i], nums[j])+dfs(mask|(1<<i)|(1<<j), k+1))
            return res
        return dfs(0, 1)

File: scripts/algorithms/M/Maximize Score After N Operations/Maximize Score After N Operations.cpp
int dp[16384];
int gcd_table[14][14];

class Solution {
public:
    int maxScore(vector<int>& nums) {
        memset(dp, -1, sizeof(dp));
        int sz = nums.size();

        // Build the GCD table 
        for (int i = 0; i < sz; ++i) {
            for (int j = i+1; j < sz; ++j) {gcd_table[i][j] = gcd(nums[i], nums[j]);}
        }

        // Looping from state 0 to (1<<sz)-1
        dp[0] = 0;
        for (int s = 0; s < (1<<sz); ++s) {
            int cnt = __builtin_popcount(s);
            if (cnt &1 )continue; // bitcount can't be odd
            for (int i = 0; i < sz; ++i) {
                if (s & (1<<i)) continue;
                for (int j = i+1; j < sz; ++j) {
                    if (s & (1<<j)) continue;
                    int next_state = s^(1<<i)^(1<<j);
                    dp[next_state] = max(dp[next_state], dp[s] + (cnt/2+1)*gcd_table[i][j]);
                }
            }
        }
        return dp[(1<<sz)-1];
    }
};